### ISSUE TITLE
Not working with Windows Server 2019 ("Cannot verify the file SHA256. Deleting the file.")

### ISSUE THREAD

**Issue by fdcastel** [MicrosoftDocs/Virtualization-Documentation#919]
Not working with Windows Server 2019:

```
C:\&gt; Install-Package -Name docker -ProviderName DockerMsftProvider

WARNING: Cannot verify the file SHA256. Deleting the file.
WARNING: C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip does not exist
Install-Package : Cannot find path 'C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip' because it does not exist.
``` 

Trying to use `-RequiredVersion 18.03.1-ee-4` gives the same error:

```
C:\&gt; Install-Package -Name docker -ProviderName DockerMSFTProvider -RequiredVersion 18.03.1-ee-4

WARNING: Cannot verify the file SHA256. Deleting the file.
WARNING: C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-03-1-ee-4.zip does not exist
Install-Package : Cannot find path 'C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-03-1-ee-4.zip' because it does not exist.
```

It seems to be a very old one:

https://github.com/OneGet/MicrosoftDockerProvider/issues/15

But, unfortunately, it appears nobody is listening:

https://github.com/MicrosoftDocs/Virtualization-Documentation/issues/917
https://github.com/MicrosoftDocs/Virtualization-Documentation/issues/873

P.S.: The date of this article (shown just below the main title) is 09/25/2016 (even being about Window Server 2019).

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 44a44120-01fd-11ca-9d51-21ac35609f11
* Version Independent ID: 75a53123-64be-1caf-fd67-7e7bef3b0aa0
* Content: [Windows Containers on Windows Server](https://docs.microsoft.com/en-us/virtualization/windowscontainers/quick-start/quick-start-windows-server#1-install-docker)
* Content Source: [virtualization/windowscontainers/quick-start/quick-start-windows-server.md](https://github.com/Microsoft/Virtualization-Documentation/blob/live/virtualization/windowscontainers/quick-start/quick-start-windows-server.md)
* Service: **windows-containers**
* Product: **windows-containers**
* GitHub Login: @cwilhit

**Comment by StefanScherer** [comment]
I can reproduce this. It sometimes happens, some time later it works again. Seems like the download bandwidth is flaky.

**Comment by Hansdevries** [comment]
Same issue here.
Busy all morning on Server 2016. but everytime error Cannot verify the file SHA256

**Comment by fdcastel** [comment]
Thanks for the info, guys but it must be another problem than just the connection. Yesterday, I left this running for **the entire day**:

```powershell
 do {
     Write-Host "Attempt #$(++$attempt; $attempt)..."
     Install-Package -Name docker -ProviderName DockerMsftProvider -Force
 } while (-not $?)
```

Today, the result:

![image](https://user-images.githubusercontent.com/629408/49332091-f2f2ea80-f58d-11e8-9d90-af193a375eb9.png)

Also, I can download the entire Server 2019 .ISO in a few minutes.

**Comment by somnio0505** [comment]
When asked package trust,  what key did you entered?
When I entered 'Y', same problem occurred. But when i entered 'A' successfully installed with SHA256 warning.

**Comment by stat91** [comment]
Try below

cd C:\Users\Administrator\AppData\Local\Temp\2\DockerMsftProvider

Start-BitsTransfer -Source https://dockermsft.blob.core.windows.net/dockercontainer/docker-18-09-0.zip -Destination docker-18-09-0.zip

Get-FileHash -Path docker-18-09-0.zip -Algorithm SHA256

Install-Package -Name docker -ProviderName DockerMsftProvider -Verbose

Restart-Computer -Force

**Comment by fdcastel** [comment]
**TL;DR: The problem occurs when using the -Force argument.** 

This combined with the fact that the package is "from a package source that is not marked as trusted".

@stat91 Thank you very much for your steps. They indeed installed Docker successfully on my system.

When running `Install-Package` without the `-Force` argument (and answering "Yes" to the question):

```
PS C:\Users\Administrator\AppData\Local\Temp\2\DockerMsftProvider> Install-Package -Name docker -ProviderName DockerMsftProvider -Verbose
VERBOSE: Using the provider 'DockerMsftProvider' for searching packages.
VERBOSE: Download size: 0.01MB
VERBOSE: Free space on the drive: 227224.14MB
VERBOSE: Downloading https://dockermsft.blob.core.windows.net/dockercontainer/DockerMsftIndex.json to C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\DockerDefault_DockerSearchIndex.json
VERBOSE: About to download
VERBOSE: Finished downloading
VERBOSE: Downloaded in 0 hours, 0 minutes, 0 seconds.
VERBOSE: Performing the operation "Install Package" on target "Package 'Docker' version '18.09.0' from 'DockerDefault'.".

The package(s) come(s) from a package source that is not marked as trusted.
Are you sure you want to install software from 'DockerDefault'?
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "N"): y
VERBOSE: Containers feature is already installed. Skipping the install.
VERBOSE: C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip already exists. Skipping save. Use -Force to overwrite.
VERBOSE: Verifying Hash of the downloaded file.
VERBOSE: Hash verified!
VERBOSE: Found C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip to install.
VERBOSE: Trying to unzip : C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip
VERBOSE: Preparing to expand...
VERBOSE: Created 'C:\Program Files\docker\docker.exe'.
VERBOSE: Created 'C:\Program Files\docker\dockerd.exe'.
VERBOSE: Created 'C:\Program Files\docker\libeay32.dll'.
VERBOSE: Created 'C:\Program Files\docker\libwinpthread-1.dll'.
VERBOSE: Created 'C:\Program Files\docker\licenses.txt'.
VERBOSE: Created 'C:\Program Files\docker\ssleay32.dll'.
VERBOSE: Created 'C:\Program Files\docker\vcruntime140.dll'.
VERBOSE: Created 'C:\Program Files\docker\docker.exe'.
VERBOSE: Created 'C:\Program Files\docker\dockerd.exe'.
VERBOSE: Trying to enable the docker service...
VERBOSE: Removing the archive: C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip

Name                           Version          Source           Summary
----                           -------          ------           -------
Docker                         18.09.0          DockerDefault    Contains Docker EE for use with Windows Server.
```



However, when running `Install-Package` with the `-Force` argument:

```
PS C:\Users\Administrator\AppData\Local\Temp\2\DockerMsftProvider> Install-Package -Name docker -ProviderName DockerMsftProvider -Verbose -Force
VERBOSE: Using the provider 'DockerMsftProvider' for searching packages.
VERBOSE: Download size: 0.01MB
VERBOSE: Free space on the drive: 227224.23MB
VERBOSE: Downloading https://dockermsft.blob.core.windows.net/dockercontainer/DockerMsftIndex.json to C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\DockerDefault_DockerSearchIndex.json
VERBOSE: About to download
VERBOSE: Finished downloading
VERBOSE: Downloaded in 0 hours, 0 minutes, 0 seconds.
VERBOSE: Performing the operation "Install Package" on target "Package 'Docker' version '18.09.0' from 'DockerDefault'.".
VERBOSE: Containers feature is already installed. Skipping the install.
VERBOSE: C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip already exists. Deleting and downloading again.
VERBOSE: Verifying Hash of the downloaded file.
WARNING: Cannot verify the file SHA256. Deleting the file.
VERBOSE: Hash verified!
WARNING: C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip does not exist
Install-Package : Cannot find path 'C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-18-09-0.zip' because it does not exist.
At line:1 char:1
+ Install-Package -Name docker -ProviderName DockerMsftProvider -Verbos ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Users\ADMINI...ker-18-09-0.zip:String) [Install-Package], Exception
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.RemoveItemCommand,Microsoft.PowerShell.PackageManagement.Cmdlets.InstallPackage
```

**Comment by dreschner** [comment]
I experience exactly the same problem when using the `-force` option. Would be great to get a fix for that as the workaround is kind of silly.

**Comment by sikansen** [comment]
Hi. I'm still having this problem. Leaving out -force parameter doesn't work.

**Comment by bilalbox** [comment]
> 
> 
> Try below
> 
> cd C:\Users\Administrator\AppData\Local\Temp\2\DockerMsftProvider
> 
> Start-BitsTransfer -Source https://dockermsft.blob.core.windows.net/dockercontainer/docker-18-09-0.zip -Destination docker-18-09-0.zip
> 
> Get-FileHash -Path docker-18-09-0.zip -Algorithm SHA256
> 
> Install-Package -Name docker -ProviderName DockerMsftProvider -Verbose
> 
> Restart-Computer -Force

Thanks @stat91 ! That worked for me as well, but I had to update the command to download the latest version (18.09.6 at the time of this posting). Here are the commands I ran:

```
 Start-BitsTransfer -Source https://dockermsft.blob.core.windows.net/dockercontainer/docker-18-09-6.zip -Destination docker-18-09-6.zip

 Get-FileHash -Path docker-18-09-6.zip -Algorithm SHA256

Install-Package -Name docker -ProviderName DockerMsftProvider -Verbose
```

**Comment by jbella** [comment]
I am running into this problem trying to download 19.03.0-rc2 using the DockerProvider.

It says the SHA256 hashes dont' match and when I download file and run the Get-FileHash on it, it indeed does not match.

The index file has this data:

```
"19.03.0-rc2": {
            "date": "2019-06-06T22:48:55.696682",
            "notes": "Docker for Windows Server 2016",
            "sha256": "254378f7ac8c9277bce60952d61ea2714829331c2e73ab0abd9d962fda98ee7d",
            "size": 130197728,
            "url": "https://download.docker.com/components/engine/windows-server/19.03/docker-19.03.0-rc2.zip"
        }
```

Running Get-FileHash on the downloaded file I get:

```
Algorithm       Hash                                                                   Path
---------       ----                                                                   ----
SHA256          572D7E7DC68044A85C13C46D3A5F81E7B677CC24C97524AE78F13BB03E86EC77       C:\Users\joebel\Downloads\docker-19.03.0-rc2.zip
```

The other versions seem to work, it's only rc2 that seems to not match.

Any idea how I can get around this? Can I tell Windows to skip the SHA verification?

**Comment by StefanScherer** [comment]
Thanks @jbella, I can reproduce your problem with the wrong sha256 sum. We are tracking this issue internally. 

But your issue seems different to the original description of this GH issue.

**Comment by rcaunt** [comment]
@StefanScherer @jbella I've just opened a fresh issue for this as we've encountered it too

https://github.com/MicrosoftDocs/Virtualization-Documentation/issues/1158

**Comment by StefanScherer** [comment]
So, the general fix seems to be to update the package provider to the latest version available.
This can be done with this command:

```powershell
Find-PackageProvider DockerMsftProvider | Install-PackageProvider
```

Then run the Install-Package command to install Docker.

**Comment by isanych** [comment]
Nope, updating provider is not helping:
```
PS C:\Users\Administrator> Find-PackageProvider DockerMsftProvider | Install-PackageProvider
PS C:\Users\Administrator> Install-Package -Name docker -ProviderName DockerMsftProvider -Force
WARNING: Cannot verify the file SHA256. Deleting the file.
WARNING: C:\Users\ADMINI~1\AppData\Local\Temp\2\DockerMsftProvider\Docker-19-03-1.zip does not exist
```
But manual download instructions above work.

**Comment by eclipses** [comment]
Downloaded the file manually worked for me

```
Start-BitsTransfer -Source https://dockermsft.blob.core.windows.net/dockercontainer/docker-19-03-1.zip -Destination C:\Users\SCASTE~1\AppData\Local\Temp\DockerMsftProvider\Docker-19-03-1.zip

Get-FileHash -Path C:\Users\SCASTE~1\AppData\Local\Temp\DockerMsftProvider\Docker-19-03-1.zip -Algorithm SHA256

Install-Package -Name docker -ProviderName DockerMsftProvider -Verbose
```

**Comment by cuichenli** [comment]
I was trying to follow the instructions above, however, when I run the command to install docker, the installer would remove the downloaded zip file. 
```
VERBOSE: Containers feature is already installed. Skipping the install.
VERBOSE: C:\Users\LOCALA~1\AppData\Local\Temp\1\DockerMsftProvider\Docker-19-03-1.zip already exists. Deleting and downloading again.
```
I am on Windows Server 2019 OS Build 17763.652. And I checked the hash value is correct. ¯\\_(ツ)_/¯

**Comment by eydunn** [comment]
@cuichenli Don't use the **-Force** flag with Install-Package

This worked for me:

```powershell
cd C:\Users\Administrator\AppData\Local\Temp\DockerMsftProvider

Start-BitsTransfer -Source https://dockermsft.blob.core.windows.net/dockercontainer/docker-19-03-1.zip 

Get-FileHash -Path C:\Users\Administrator\AppData\Local\Temp\DockerMsftProvider\Docker-19-03-1.zip -Algorithm SHA256

Install-Package -Name docker -ProviderName DockerMsftProvider -Verbose

Start-Service Docker
```

**Comment by cuichenli** [comment]
@eydunn Yes this will work, thanks! But how to bypass the confirm? Want to run this in automation scripts.

**Comment by petemidge** [comment]
Downloading the file manually worked for me once I put the proper path in the destination flag:
`Start-BitsTransfer -Source https://dockermsft.blob.core.windows.net/dockercontainer/docker-19-03-1.zip -Destination C:\Users\Administrator\AppData\Local\Temp\2\DockerMsftProvider\docker-19-03-1.zip`

**Comment by bctails** [comment]
We're all fiddling here, I'm trying to install a Docker machine for a friend and it's just not working.
Tried above script with seperately downloading, also tried without -Force, but it's not coming through.

Does anyone actually why this is happening?

**Comment by bctails** [comment]
Just reporting back here. In our case, manually selecting an older version of Docker (18.09) seems to have solved the problem.

**Comment by Rostis1av** [comment]
https://github.com/Rostis1av/unstable/blob/master/docker.ps1

me decided it by married MS & docker official info

**Comment by Rostis1av** [comment]
https://github.com/Rostis1av/unstable/blob/master/docker-install.ps1

**Comment by bchavez** [comment]
```powershell
Install-WindowsFeature -Name Containers
Install-Module -Name DockerMsftProvider -Repository PSGallery -Force
Install-Package -Name docker -ProviderName DockerMsftProvider -Force -RequiredVersion 18.09.8 
Start-Service docker
docker --version
```
```
Docker version 18.09.8, build 90e30bdf98
```

**Comment by collabnix** [comment]
@bchavez You missed out a command before Start-Service docker

```
(Install-WindowsFeature Containers).RestartNeeded
```

OR 

You need to restart the Windows Server.

**Comment by nonbob** [comment]
I have 1.0.0.8 of DockerMsftProvider and am still seeing this.  Anybody else?

**Comment by Cinezi** [comment]
> ```powershell
> Install-WindowsFeature -Name Containers
> Install-Module -Name DockerMsftProvider -Repository PSGallery -Force
> Install-Package -Name docker -ProviderName DockerMsftProvider -Force -RequiredVersion 18.09.8 
> Start-Service docker
> docker --version
> ```
> 
> ```
> Docker version 18.09.8, build 90e30bdf98
> ```

Thanks !
Works for me !

**Comment by mukhtar2t** [comment]
# Trust Package Source
Set-PackageSource -Name DockerDefault -NewName DockerDefault2 -ProviderName DockerMsftProvider -Trusted
Set-PackageSource -Name DockerDefault2 -NewName DockerDefault -ProviderName DockerMsftProvider -Trusted
Find-Package -ProviderName DockerMsftProvider | Install-Package -Verbose
